pkg/threescale/connectors/backend.go

```diff
@@ -0,0 +1,141 @@
+package connectors
+
+import (
+	"fmt"
+	"time"
+
+	backend "github.com/3scale/3scale-go-client/client"
+	"github.com/3scale/3scale-istio-adapter/pkg/threescale/metrics"
+)
+
+type AuthRepResult string
+
+const Unknown AuthRepResult = "UNKNOWN"
+const Ok AuthRepResult = "AUTHENTICATED"
+const Unauthenticated AuthRepResult = "UNAUTHENTICATED"
+const cacheMiss AuthRepResult = "UNCACHED"
+
+const (
+	UserKey = iota + 1
+	AppId
+)
+
+type authRepFn func(auth backend.TokenAuth, key string, svcID string, params backend.AuthRepParams, extension map[string]string) (backend.ApiResponse, error)
+
+type AuthRepUserAuth struct {
+	// Api Key Authn Pattern
+	UserKey string
+	// App Id Authn Pattern - Mutually exclusive with Api Key Pattern
+	AppID string
+	// Optional App Key as a key pair for App ID pattern
+	AppKey   string
+	AuthType int
+}
+
+type AuthRepRequest struct {
+	ServiceID string
+	UserAuth  AuthRepUserAuth
+	TokenAuth backend.TokenAuth
+	Metrics   backend.Metrics
+}
+
+type authRepRemoteCall struct {
+	authKey        string
+	params         backend.AuthRepParams
+	callback       authRepFn
+	remoteHostname string
+}
+
+type BackendCache interface {
+	AuthRep(svcID string, metrics backend.Metrics) AuthRepResult
+	Set()
+}
+
+type BackendConnector struct {
+	cache           BackendCache
+	reportMetricsFn metrics.ReportMetricsFn
+}
+
+type LocalCache struct {
+}
+
+func (l LocalCache) AuthRep(svcID string, metrics backend.Metrics) AuthRepResult {
+	return cacheMiss
+}
+
+func (l LocalCache) Set() {
+
+}
+
+func NewBackendConnector() *BackendConnector {
+	return &BackendConnector{}
+}
+
+func (bc *BackendConnector) DoAuthRep(req AuthRepRequest, metrics backend.Metrics, c *backend.ThreeScaleClient) (backend.ApiResponse, error) {
+	var resp backend.ApiResponse
+
+	callRemote := func() (backend.ApiResponse, error) {
+		remoteReq, err := bc.authRepRemoteReqBuilder(req.UserAuth, metrics, c)
+		if err != nil {
+			return resp, err
+		}
+
+		return bc.callRemote(req.TokenAuth, remoteReq, req.ServiceID)
+	}
+
+	if bc.cache != nil {
+		switch result := bc.cache.AuthRep("", nil); result {
+		case cacheMiss:
+			resp, err := callRemote()
+			if err != nil {
+				bc.cache.Set()
+			}
+			return resp, err
+		case Unauthenticated:
+			resp.Success = false
+			resp.Reason = "limits exceeded"
+			resp.StatusCode = 429
```

Comment:
It's also important to think about the unauth reason, because the client needs to know to lower the rate (429 Too Many Requests) or whether it got something wrong, that is, `limits exceeded` is one option, but there are a few more (see `3scale-rejection-reason` header and extension, IIRC).

Reply:
Right, my thinking was that if we have are using caching and we have a cache hit then it would be the only rejection reason as we wouldn't be going back to 3scale at that point?

Reply:
We want to cache uanthorized statuses caused by other reasons as well.

In particular, imagine someone misconfigures their client, and uses the wrong `app_id` (or `service_id`, or ...). Now imagine that someone is calling us 10k times per second. We certainly don't want to call 3scale 10k times/s for a combination of service and app that we know is faulty.

So caching other causes for unauth is also important, if only to prevent the case depicted above.

Reply:
Yes, makes sense. Thanks